### PR TITLE
WIP FH-3065 Add mongodb impl for the Updates model for sync

### PR DESCRIPTION
**Pull request by david-martin**
* abstracts the '<dataset>-updates' fhdb model updates to a separate model file.
* adds a second update model that connects with the mongo client rather than fhdb
  * this allows for doing a create and delete without having to do a list first (fhdb limitations)

### CODE REVIEW

File: lib/sync-UpdatesModel_fhdb.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        return cb(null, data);
+      });
+    }
+    else {
+      return cb(null, data);
+    }
+  });
+};
+
+module.exports = {
```
Comment (fh-qube) [inline lib/sync-UpdatesModel_fhdb.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/sync-UpdatesModel_fhdb.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    "type": dataset_id + "-updates",
+    "eq": {
+      "cuid": data.cuid,
+      "type": data.type,
+      "action": data.action,
+      "hash": data.hash,
+      "uid": data.uid
+    }
+  }, function (err, found) {
+    if (err) return cb(err);
```
Comment (fh-qube) [inline lib/sync-UpdatesModel_fhdb.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync-UpdatesModel_fhdb.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  // Check if the record already exists, then remove it if it does
+  fhdb({
+    "act": "list",
+    "type": dataset_id + "-updates",
+    "eq": {
+      "cuid": data.cuid,
+      "hash": data.hash
+    }
+  }, function (err, res) {
+    if (err) return cb(err, data);
```
Comment (fh-qube) [inline lib/sync-UpdatesModel_fhdb.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync-UpdatesModel_fhdb.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  });
+};
+
+var list = function(dataset_id, eq, cb) {
+  fhdb({
+    "act": "list",
+    "type": dataset_id + "-updates",
+    "eq": eq
+  }, function(err, records) {
+    if (err) return cb(err);
```
Comment (fh-qube) [inline lib/sync-UpdatesModel_fhdb.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync-UpdatesModel_fhdb.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    if (res && res.list && res.list.length > 0) {
+      var rec = res.list[0];
+      var uid = rec.guid;
+      fhdb({
+        "act": "delete",
+        "type": dataset_id + "-updates",
+        "guid": uid
+      }, function (err) {
+        if (err) return cb(err, data);
```
Comment (fh-qube) [inline lib/sync-UpdatesModel_fhdb.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync-UpdatesModel_mongo.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+var remove = function(dataset_id, data, cb) {
+  var toDelete = {
+    cuid: data.cuid,
+    hash: data.hash
+  };
+  return mongo.collection(dataset_id + "-updates").deleteOne(toDelete, cb);
+};
+
+module.exports = {
```
Comment (fh-qube) [inline lib/sync-UpdatesModel_mongo.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/sync-UpdatesModel_mongo.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+var fhdb;
+var mongo;
+
+var setFHDB = function(db) {
+  fhdb = db;
+};
+
+var setConnectionUrl = function(url) {
+  MongoClient.connect(url, function(err, db) {
+    if (err) throw err;
```
Comment (fh-qube) [inline lib/sync-UpdatesModel_mongo.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

**Comment by fh-qube** [comment]
SonarQube analysis reported 7 issues:
* BLOCKER 2 blocker
* MINOR 5 minor

Watch the comments in this conversation to review them.

File: package.json [hunk 8]
```diff
@@ -16,6 +16,7 @@
     "lodash-contrib": "^393.0.1",
     "memcached": "^2.0.0",
     "moment": "2.14.1",
+    "mongodb": "^2.1.18",
```
Comment (danbev) [inline package.json]: Do we need a run of npm shrinkwrap for this too?

**Comment by fh-qube** [comment]
SonarQube analysis reported 7 issues:
* BLOCKER 2 blocker
* MINOR 5 minor

Watch the comments in this conversation to review them.

**Comment by fh-qube** [comment]
SonarQube analysis reported 7 issues:
* BLOCKER 2 blocker
* MINOR 5 minor

Watch the comments in this conversation to review them.

**Comment by danbev** [comment]
I've rebased this with upstream master. Hope I've not caused problems for anyone by doing this.

**Comment by fh-qube** [comment]
SonarQube analysis reported 5 issues:
* BLOCKER 1 blocker
* MINOR 4 minor

Watch the comments in this conversation to review them.

File: lib/sync-UpdatesModel_mongo.js (hunk 6 above)
Comment (fh-qube) [inline lib/sync-UpdatesModel_mongo.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

File: lib/sync-UpdatesModel_mongo.js (hunk 7 above)
Comment (fh-qube) [inline lib/sync-UpdatesModel_mongo.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

**Comment by fh-qube** [comment]
SonarQube analysis reported 7 issues:
* BLOCKER 2 blocker
* MINOR 5 minor

Watch the comments in this conversation to review them.

**Comment by fh-qube** [comment]
SonarQube analysis reported 5 issues:
* BLOCKER 1 blocker
* MINOR 4 minor

Watch the comments in this conversation to review them.

**Comment by danbev** [comment]
Not really sure about the build failure reported by my lastest update. Re-running the build [succeeds](http://bob.feedhenry.net:8080/job/fh-mbaas-api_pullrequests/523/console)

**Comment by aidenkeating** [comment]
@danbev Tried using the updated `FH-3065-save-sync-updates-with-mongo-create` branch with testing cloud app and I don't seem to be able to install it as a dependency due to the same issue seen in the build. Taking a closer look now.

**Comment by danbev** [comment]
@aidenkeating What version of node are you using?

**Comment by david-martin** [comment]
Closing this WIP as it has served its purpose as a POC

**Pull request closed without merging**